pkg/controller/direct/tpu/tpuvirtualmachine_controller.go

```diff
@@ -0,0 +1,338 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tpu
+
+// +tool:controller
+// proto.service: google.cloud.tpu.v2.Tpu
+// proto.message: google.cloud.tpu.v2.Node
+// crd.type: TPUVirtualMachine
+// crd.version: v1alpha1
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/googleapis/gax-go/v2"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/common"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+
+	pb "cloud.google.com/go/tpu/apiv2/tpupb"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/tpu/v1alpha1"
+	kcc "github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/k8s/v1alpha1"
+)
+
+func init() {
+	registry.RegisterModel(krm.TPUVirtualMachineGVK, NewTPUVirtualMachineModel)
+}
+
+func NewTPUVirtualMachineModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &modelTPUVirtualMachine{config: config}, nil
+}
+
+var _ directbase.Model = &modelTPUVirtualMachine{}
+
+type modelTPUVirtualMachine struct {
+	config *config.ControllerConfig
+}
+
+func (m *modelTPUVirtualMachine) AdapterForObject(ctx context.Context, kube client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.TPUVirtualMachine{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	i, err := obj.GetIdentity(ctx, kube)
+	if err != nil {
+		return nil, err
+	}
+	id := i.(*krm.TPUVirtualMachineIdentity)
+	project := &refs.Project{ProjectID: id.ParentID.ProjectID}
+
+	if err := common.NormalizeReferences(ctx, kube, obj, project); err != nil {
+		return nil, err
+	}
+
+	gcpClient, err := newGCPClient(ctx, m.config)
+	if err != nil {
+		return nil, err
+	}
+	tpuClient, err := gcpClient.newClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desiredProto := TPUVirtualMachineSpec_ToProto(mapCtx, &obj.Spec)
```

Comment:
Suggest moving this call to Create and Update to preserve the actual KRM value for some key information like defaulting, unset/unmanaged, and "OUTPUT-ONLY" name that is required in Create.